Guaranteed TCO (#3407)
rust-lang/rfcs: text/0000-explicit-tail-calls.md

<!-- If applicable, provide sample error messages, deprecation warnings, or migration guidance. -->
<!-- (TODO Error messages once an initial implementation exists) -->

<!-- (TODO migration guidance) -->

## Teaching
For new Rust programmers this feature should probably be introduced late into the learning process, it requires
understanding some advanced concepts and the current use cases are likely to be niche. So it should be taught similarly
as to programmers that already know Rust.

## Examples
On to some examples. Starting with how `return` and `become` differ, two example use cases, and some potential
pitfalls. 

### The difference between `return` and `become`
[difference]: #difference
The difference to `return` is that `become` drops function local variables **before** the function call
instead of after. So the following function ([original example](https://github.com/rust-lang/rfcs/issues/2691#issuecomment-1136728427)):
```rust
fn x() {
    let a = Box::new(());
    let b = Box::new(());
    become y(a);
}
```

The drops will be elaborated by the compiler like this:
```rust
fn x() {
    let a = Box::new(());
    let b = Box::new(());

>     drop(b); // `a` is not dropped because it is moved to the callee

Reviewer: It's unclear to me from the text whether the drop will happen before evaluation of function arguments or not.

Say we have this:
```rust
fn x() {
    let a = Box::new(());
    let b = Box::new(());
    become y(foo(a));
}
```
whether it will be elaborated to
```rust
fn x() {
    let a = Box::new(());
    let b = Box::new(());
    drop(b);
    become y(foo(a));
}
```
or
```rust
fn x() {
    let a = Box::new(());
    let b = Box::new(());
    let arg0 = foo(a);
    drop(b);
    become y(arg0);
}
```
?

From the literal reading of this text it sounds to me that the implication is that drop happens before evaluation of function arguments. And I think this can be very problematic for drop elaboration, as we currently don't have any case where whether a variable will be dropped depends on a later use.

I think we should instead clearly specify the order of evaluation -- the function arguments of the tail call happens first, then it's deemed as end-of-scope of the function, then the tail function call happens.

So essentially we can treat it similar to this alternative function:
```rust
fn x() {
    let (arg0,)= {
        let a = Box::new(());
        let b = Box::new(());
        (foo(a), );
    };
    become y(arg0)
}
```

Author: Good point! I'll update the text as you suggest. Though, I'm not sure showing this with a scope is the right way, as it can't really be translated if the `become` is used inside an existing scope. Would the following example be clear enough?

```rust
fn x(_arg: ()) {
    let a = Box::new(());
    let b = Box::new(());

    // Move to temporary variables.
    let function_ptr = bar; // The function pointer could be the result of an expression like: fn_list[fn_idx];
    let arg0 = calc(a);

    // End of the function, all variables not used in the `become` call are dropped, as would be done after a `return`.
    drop(b);
    // `a` has been moved into `calc` and does not need to be dropped.
    drop(_arg);

    // Finally, `become` the called function.
    become function_ptr(arg0);
}
```

Author: Just for reference, there is also a more detailed section on borrowchecking: https://github.com/phi-go/rfcs/blob/guaranteed-tco/text/0000-explicit-tail-calls.md#borrowchecking

Reviewer: For arbitrarily complex functions, this desugaring would work:
```rust
fn x() {
    let a = Box::new(());
    let b = Box::new(1);
    if C {
        let c = Box::new(());
        become y(foo(a));
    }
    if D {
        become z(bar(b));
    }
    // return ();
}
```
to
```rust
enum XReturn {
    Become1(Box<()>),
    Become2(Box<i32>),
    Return(()),
}
fn x_inner() {
    let a = Box::new(());
    let b = Box::new(1);
    if C {
        let c = Box::new(());
        return XReturn::Become1(foo(a));
    }
    if D {
        return XReturn::Become2(bar(b));
    }
    return XReturn::Return(());
}
fn x() {
    match x_inner() {
        XReturn::Become1(arg0) => become y(arg0),
        XReturn::Become2(arg0) => become z(arg0),
        XReturn::Return(ret) => return ret,
    }
}
```
Essentially [defunctionalization](https://en.wikipedia.org/wiki/Defunctionalization).

Author: Thank you for the example, this is an approach I did not think of. Though, as the original example is part of the guide-level explanation, I'm not quite convinced that using the scope example is the best way to teach about the performed drop elaboration. For the implementation there is also the section on borrowchecking which is hopefully already precise enough.

Author: I have updated the section by merging it with the borrow checking section, see here: https://github.com/rust-lang/rfcs/pull/3407/commits/0b47449cb4b5ae6b4ecbb7ef19c4e6da650e8b05.

As this feature is strictly opt-in and the `become` keyword is already reserved, this has no impact on existing code.

<!-- If applicable, provide sample error messages, deprecation warnings, or migration guidance. -->
<!-- (TODO Error messages once an initial implementation exists) -->

<!-- (TODO migration guidance) -->

## Teaching
For new Rust programmers this feature should probably be introduced late into the learning process, it requires
understanding some advanced concepts and the current use cases are likely to be niche. So it should be taught similarly
as to programmers that already know Rust.

## Examples
On to some examples. Starting with how `return` and `become` differ, two example use cases, and some potential
pitfalls. 

### The difference between `return` and `become`
[difference]: #difference
The difference to `return` is that `become` drops function local variables **before** the `become` function call
instead of after. To be more specific a `become` expression acts as if the following events occurred in-order:

> 1. All variables that are being passed by-value are moved to temporary storage.

Reviewer: ```suggestion
1. Function call arguments are evaluated into temporary storage. If a local variable is used as a value in the arguments, it is moved.
```

Author: That is clearer, thank you!

- Feature Name: explicit_tail_calls
- Start Date: 2023-04-01
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

> While tail call elimination (TCE) is already possible via tail call optimization (TCO) in Rust, there is no way to guarantee that a stack frame should be reused.

Author: updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/94916aa0540e545ef89cd95c86821f849ce1fad4

This RFC describes a language feature providing tail call elimination via the `become` keyword providing this guarantee.
If this guarantee can not be provided by the compiler a compile time error is generated instead.

# Motivation
[motivation]: #motivation
Tail call elimination (TCE) allows stack frames to be reused.

> While TCE via tail call optimization (TCO) is already supported by Rust, as is normal for optimizations TCE will only be applied if the compiler expects an improvement by doing so.

Author: updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/94916aa0540e545ef89cd95c86821f849ce1fad4

> There is currently no way to specify that TCE should be guaranteed.

Author: updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/94916aa0540e545ef89cd95c86821f849ce1fad4

- If applicable, provide sample error messages, deprecation warnings, or migration guidance.
- If applicable, describe the differences between teaching this to existing Rust programmers and new Rust programmers.
- Discuss how this impacts the ability to read, understand, and maintain Rust code. Code is read and modified far more often than written; will the proposed feature make code easier to maintain?

For implementation-oriented RFCs (e.g. for compiler internals), this section should focus on how compiler contributors should think about the change, and give examples of its concrete impact. For policy RFCs, this section should provide an example-driven introduction to the policy, and explain its impact in concrete terms. -->
Pretending this RFC has already been accepted into Rust, it could be explained to another Rust programmer as follows.

## Tail Call Elimination
[tail-call-elimination]: #tail-call-elimination

Rust supports a way to guarantee tail call elimination (TCE) for function calls using the `become` keyword.

> If TCE is requested for a call the called function will reuse the stack frame of the calling function, assuming all requirements are fulfilled.

Author: updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/ae758f560cda30d83d4d6b6294bc1ac4bc333455

Note that TCE can opportunistically also be performed by Rust using tail call optimization (TCO), this will cause TCE to be used if it is deemed to be "better" (as in faster, or smaller if optimizing for space).

TCE is interesting for two groups of programmers: Those that want to use recursive algorithms,
which can overflow the stack if the stack frame is not reused; and those that want to create highly optimized code
as creating new stack frames can be expensive.

To request TCE the `become` keyword can be used instead of `return`.
Note that, as both keywords act as the end of the function,
`become` can be used everywhere that `return` is used.

> However, there are several requirements on the called function which need to be fulfilled for TCE (and TCO) to work.

Author: updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/ae758f560cda30d83d4d6b6294bc1ac4bc333455

The main restriction is that the argument to `become` is a tail call,
a call that is the last action performed in the function.
Supported are calls such as `become foo()`, `become foo(a)`, `become foo(a, b)`, `become foo(1 + 1)`,
`become foo(bar())`, `become foo.method()`, or `become function_table[idx](arg)`.
Calls that are not in the tail position can **not** be used, for example, `become foo() + 1` is not allowed.
In the example, the function would need to be evaluated and **then** the addition would need to take place.

A further restriction is on the function signature of the caller and callee.
The stack frame layout is based on the calling convention, arguments, as well as return types (the function signature in
short).
As the stack frame is to be reused it needs to be similar enough for both functions.
This requires that the function signature and calling convention of the calling and called function need to match exactly.

Additionally, there is a further restriction on the arguments.
The stack frame of the calling function is reused, it is essentially cleaned up and the called function takes the space.

> As a result, it is not possible to pass references to local variables, nor will the called function "return" to the calling function. So all variables not used as an argument are dropped before the call and no cleanup will be done after the call.

Author: updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/ae758f560cda30d83d4d6b6294bc1ac4bc333455

        accum += data[offset];
        offset += 1;
        become sum_list(data, offset, accum); // <- become here
    } else {
        accum // <- equivalent to `return accum;`
    }
}
```

### Use Case 2: Interpreter
For an interpreter, the usual loop is to get an instruction, match on that instruction to find the corresponding function, **call** that function, and finally return to the loop to get the next instruction. (This is a simplified example.)

```rust
fn exec_instruction(mut self) {
    loop {
        let next_instruction = self.read_instr();
        match next_instruction {
            Instruction::Foo => self.execute_instruction_foo(),
            Instruction::Bar => self.execute_instruction_bar(),
        }
    }
}
```

This example can be turned into the following code, which no longer does any calls and instead just uses jump instructions. (Note that this example might not be the optimal way to use `become`.)

```rust
fn execute_instruction_foo(mut self) {
    // foo things ...

    become self.next_instruction();
}

fn execute_instruction_bar(mut self) {
    // bar things ...

    become self.next_instruction();
}

fn next_instruction(mut self) {

>     let next_instruction = self.read_instr(); // this call can be inlined

Reviewer 2: I don't think this is necessary at all.

In any case, these can still be called as methods. One example of their error-prone nature ([source](https://github.com/rust-lang/rfcs/pull/3407#discussion_r1167112296)):
```rust
pub fn fibonacci(n: u64) -> u64 {
    if n < 2 {
        return n
    }
    become fibonacci(n - 2) + fibonacci(n - 1)
}
```
In this case, a naive author might assume that this is going to be a stack space-efficient implementation since it uses tail recursion instead of normal recursion. However, the outcome is more or less the same since the critical recursive calls are not actually in tail call position.

Further confusion could result from the same-signature restriction where the Rust compiler raises an error since fibonacci and <u64 as Add>::add do not share a common signature.

# Drawbacks
[drawbacks]: #drawbacks
<!-- Why should we *not* do this? -->
As this feature should be mostly independent of other features the main drawback lies in the implementation and
maintenance effort. This feature adds a new keyword which will need to be implemented not only in Rust but also in other
tooling. The primary effort, however, lies in supporting this feature in the backends:

> - LLVM supports a `musttail` marker to indicate that TCE should be performed [docs](https://llvm.org/docs/LangRef.html#id327). Clang which already depends on this feature seems to only generate correct code for the x86 backend [source](https://github.com/rust-lang/rfcs/issues/2691#issuecomment-1490009983) (as of 30.03.23).

Reviewer 3: ```suggestion
- LLVM supports a `musttail` marker to indicate that TCE should be performed [docs](https://llvm.org/docs/LangRef.html#id327). Clang which already depends on this feature seems to only generate correct code for the x86 backend [source](https://github.com/rust-lang/rfcs/issues/2691#issuecomment-1490009983) (as of 2023-03-30).
```

- GCC does seem to support an equivalent `musttail` marker, though it is only accessible via the [libgccjit API](https://gcc.gnu.org/onlinedocs/gcc-7.3.0/jit/topics/expressions.html#gcc_jit_rvalue_set_bool_require_tail_call) ([source](https://github.com/rust-lang/rfcs/pull/3407#discussion_r1160013809)).

> - WebAssembly accepted tail calls into the [standard](https://github.com/WebAssembly/proposals/pull/157/) and Cranelift is now [working](https://github.com/bytecodealliance/rfcs/pull/29) towards supporting it.

Reviewer 4: None of the wasm targets enable the tail call feature and they likely can't enable it for a long time due to back compat reasons.

Reviewer 4: Also Cranelift will only support tail calls for the ABI used for Webassembly. It will not support it for the SystemV or WindowsFastcall ABI's used for rust code.

Reviewer 5: > None of the wasm targets enable the tail call feature and they likely can't enable it for a long time due to back compat reasons.

As of now `tail-calls` are not yet part of the WebAssembly standard so I'd assume that there will at least be the option to generate them even though this option will probably be disabled by default. That's the same with other post-MVP Wasm proposals from what I know. @bjorn3 Please correct me if I wrong about this assumption?

> Also Cranelift will only support tail calls for the ABI used for Webassembly. It will not support it for the SystemV or WindowsFastcall ABI's used for rust code.

Does this mean that Rust code needs something like `extern "Wasm"` applied to functions for tail calls to work under Cranelift? I am confused. 🤔

Your points should not be a blocker for this RFC to be accepted but they might be blocker for the stabilization of it later on.

Reviewer 4: > As of now tail-calls are not yet part of the WebAssembly standard but proposed to be so I'd assume that there will at least be the option to generate them for producers even though this option will probably be disabled by default. That's the same with other post-MVP Wasm proposals from what I know. @bjorn3 Please correct me if I wrong about this assumption?

Yeah, I did assume LLVM has a target feature to enable it, however rustc can't enable this target feature for wasm32-unknown-unknown as it did break running rustc compiled wasm modules on wasm runtimes that don't support tail calls. This likely happens even if the user didn't explicitly wrote a tail call as LLVM will likely optimize some calls into tail calls.

> Does this mean that Rust code needs something like extern "Wasm" applied to functions for tail calls to work under Cranelift? I am confused. thinking

Cranelift supports multiple calling conventions like SystemV and WindowsFastcall. These are the default calling conventions for C on Unix and Windows respectively. The Rust ABI used by the LLVM backend is based on the same calling convention as used by C. To remain ABI compatible, cg_clif uses the same calling convention as the LLVM backend. Cranelift will not support tail calls for the SystemV and WindowsFastcall calling conventions. Instead it will only support it for the Cranelift specific Tail calling convention designed specifically for use in Webassembly runtimes. This calling convention is not supported by LLVM, so we can neither switch the default Rust ABI to it, not expose it as `extern "Tail"` in Rust. In other words cg_clif can't remain ABI compatible with cg_llvm and support explicit tail calls. Furthermore the Tail calling convention exclusively uses the stack for passing arguments, making it slower, but also much easier to implement tail calls.

Author: Thank you for the info @bjorn3, I'll add a link to this discussion in the RFC.

Reviewer: For the signature matching case we can do defunctionalization and TCE in MIR level.

Reviewer 4: Based on https://github.com/bytecodealliance/wasmtime/issues/1065#issuecomment-1367015529 I think it isn't possible to implement in the general case for the x86_64 SystemV calling convention.

> The general problem to wrestle with is easiest to see (IMHO) in a minimized example: f and g tail-call each other, both have enough arguments that some of them are passed on the stack, and they have different numbers of arguments. E.g., on x86-64 where the first six i64 args are passed in registers, say that f has 8 i64 args and g has 10 i64 args.
>
> Then when f tail-calls g, it needs 32 bytes of stack-arg space (the remaining 4 i64 args), but it itself only was given 16 bytes of stack-arg space. Furthermore, the original caller of f, in the standard SysV ABI, is responsible for popping those 16 bytes of space. There's no way to satisfy all the constraints for a true tail-call in this case in the SysV ABI. That's why the usual way to achieve true tail-calls in the arbitrary case is to adopt a "callee-pop" ABI instead: f and g both expect a certain number of args on the stack, and pop them before returning. This way, f's tail-call to g can do that pop, and then push g's args; and when g returns to the original caller of f, there's no need to somehow know that the actual returnee had 32 bytes rather than 16 of stack args.

@cfallin Do I understand this correctly?

Reviewer 6: If you restrict tail calls to functions with the exact same signature then all you need to do is overwrite the arguments of the current function in-place with new values.

Reviewer 7: > Based on [bytecodealliance/wasmtime#1065 (comment)](https://github.com/bytecodealliance/wasmtime/issues/1065#issuecomment-1367015529) I think it isn't possible to implement in the general case for the x86_64 SystemV calling convention.
> 
> > The general problem to wrestle with is easiest to see (IMHO) in a minimized example: f and g tail-call each other, both have enough arguments that some of them are passed on the stack, and they have different numbers of arguments. E.g., on x86-64 where the first six i64 args are passed in registers, say that f has 8 i64 args and g has 10 i64 args.
> > Then when f tail-calls g, it needs 32 bytes of stack-arg space (the remaining 4 i64 args), but it itself only was given 16 bytes of stack-arg space. Furthermore, the original caller of f, in the standard SysV ABI, is responsible for popping those 16 bytes of space. There's no way to satisfy all the constraints for a true tail-call in this case in the SysV ABI. That's why the usual way to achieve true tail-calls in the arbitrary case is to adopt a "callee-pop" ABI instead: f and g both expect a certain number of args on the stack, and pop them before returning. This way, f's tail-call to g can do that pop, and then push g's args; and when g returns to the original caller of f, there's no need to somehow know that the actual returnee had 32 bytes rather than 16 of stack args.
> 
> @cfallin Do I understand this correctly?

This can be worked around by using a trampoline.  f and g gain an additional boolean argument that indicates if they were called by the trampoline or not.  If they were, then true is passed, and tail call is performed by returning to the trampoline, which then pops everything and makes the next call.  Non-tail calls are made to a wrapper that passes false.  The extra parameter can be stored in a caller-save scratch register on ABIs that have that.

Reviewer 4: > That said, the extern "Rust" ABI is explicitly unspecified, so Rust could choose to use a callee-pops calling convention to make arbitrary tail calls fast.

There are several calling conventions in LLVM that support tail calls, but none of them are usable for rustc's purpose on all platforms:

* "fastcc"/"tailcc": AFAIK this one is not stable across LLVM version, which is required to keep other backends ABI compatible. I'm also not sure if it is allowed for functions exported from the current cgu.
* "cc 10" (GHC call conv): Only supports x86 and x86_64
* "cc 11" (HiPE call conv): Only supports x86 and x86_64
* "swifttailcc": x86/x86_64 and aarch64 only I think

Also the only targets on which LLVM supports tail calls are x86/x86_64 (only linkage unit local in case of PIC code like we do by default), PowerPC (only linkage unit local functions), AArch64, and WebAssembly.

Reviewer 4: As for the trampoline option, that would work I guess, but is slower and would complicate all backends to support it.

Reviewer 8: > This can be worked around by using a trampoline. f and g gain an additional boolean argument that indicates if they were called by the trampoline or not. If they were, then true is passed, and tail call is performed by returning to the trampoline, which then pops everything and makes the next call. Non-tail calls are made to a wrapper that passes false. The extra parameter can be stored in a caller-save scratch register on ABIs that have that.

Indeed, but now you've invented a calling convention that isn't quite SysV. The appeal of a "true" SysV tail-call solution is that a caller can call a callee and *not* set up this extra parameter, and have no knowledge of the tail-call convention, and everything would still work.

I'm also generally not too fond of trampoline-based approaches: they exert a pressure on users of tail calls and discourage their use in general. If we're going to have tail calls, it's preferable that they're as efficient or almost as efficient as ordinary calls -- that will unlock the truly interesting use-cases (languages unconditionally using tail-call instructions in tail positions to give Scheme-like guarantees; compiling every basic block to a little function and tail-calling to work around structured control flow limitations; using tail calls for ICs (inline caches); and maybe others).

Reviewer 7: > If we're going to have tail calls, it's preferable that they're as efficient or almost as efficient as ordinary calls

I agree, but that will require a callee-pops calling convention.

Author: @nbdd0121 correct me if I'm wrong but defunctionalization will only work for statically known function calls, not dynamic calls or calls to external functions.

@bjorn3 and @cfallin what are your thoughts on supporting tail calls only between functions with matching function signatures? As @Amanieu said, supporting tail calls with this restriction should be possible independent of the calling convention.

Reviewer 8: @phi-go I think supporting this restriction at least is still a very nice and usable building-block!

Author: Updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/b6b309492602bf0a5b1733022234110ca2678632

Let me know if you disagree with the wording or want another change.

Additionally, this proposal is limited to exactly matching function signatures which will *not* allow general tail-calls, however, the work towards this initial version is likely to be useful for a more comprehensive version.

There is also an unwanted interaction between TCE and debugging. As TCE by design elides stack frames this information is lost during debugging, that is the parent functions and their local variable values are incomplete. As TCE provides a semantic guarantee of constant stack usage it is also not generally possible to disable TCE for debugging builds as then the stack could overflow. (Still, maybe a compiler flag could be provided to temporarily disable TCE for debugging builds. As suggested [here](https://github.com/rust-lang/rfcs/pull/3407/files#r1159817279), another option would be special support for `become` by a debugger. With this support the debugger would keep track of the N most recent calls providing at least some context to the bug.)

> # Rationale and alternatives

Reviewer 9: I'd love to see more about the backend or not constraint here.

Is there a version of this that's slightly more restricted but would allow us to reasonably be able to do this even *without* backend support?

What are the code patterns of this where the tail call is essential?  Are tail calls needed between different crates?  Or in practice do people only *need* tail calls between things inside the same crate, say?

Reviewer 5: > I'd love to see more about the backend or not constraint here.

@scottmcm Can you please describe what you do mean with "backend" and "not contraint" here?

> What are the code patterns of this where the tail call is essential?

Concerning the downside of debugging with tail calls versus regular calls I'd say that tail calls behave very similar to a common iteration approach and therefore a tail call based algorithm is just as well or as badly debuggable as an iterative version of the same algorithm. Given this argumentation and that most efficient algorithms in today Rust are iterative algorithms I don't see this as a big problem personally. Please correct me if I am wrong.

> Are tail calls needed between different crates?

That's a very interesting question. For my personal use cases I have none which require tail calls across crates. However, what exactly would we win by having a constraint to only allow tail calls within the same crate if that's why you were asking?

Author: > Can you please describe what you do mean with "backend" and "not contraint" here?

I think the sentence is meant in the sense of should backends be required (as in should there be a constraint on them) to support the tail calls or not. A backend would be what generates machine code given the MIR, so for example LLVM or Cranelift.

Let me try to answer that question and I'd love to be corrected if I'm misunderstanding something.

So, I can come up with five approaches to support tail calls with various tradeoffs (and requirements on the backends), from least to most comprehensive:

1. Use an approach like defunctionalization in the MIR to effectively do trampolining without any backend needing to know that this has been done. This approach has the restriction that it only support *static* function calls. Sadly, *dynamic* function calls are where tail calls get really interesting ([example](https://github.com/rust-lang/rfcs/pull/3407#issuecomment-1500291721)). This approach can also be done for intra-crate functions by using macros, I think there exists a library for this approach already.
2. Do analysis in the MIR to try and also support dynamic calls. From what I read, this seems to be a task too difficult to do and would not support tail calls to external functions.
3. Require as little as possible from backends, this is the approach chosen in the RFC. By restricting function arguments to exactly match it should be possible for backends to support tail calls regardless of the calling convention, see for example this [discussion](https://github.com/rust-lang/rfcs/pull/3407#discussion_r1186003262). While general tail calls are hard to support, this restriction seems more feasible, though I would be interested how feasible it really is. Note that while this approach does not support general tail calls, I would say this is still quite flexible. It also supports all calls from static, dynamic and even external function calls.
4. (Just for completeness, I know this moves into the other direction of "without backend support".) One other option is to mark all functions (not function calls) to use a backend specific calling convention that supports tail calls, this obviously requires backend support. Marking the functions is done as the default calling convention usually does not support tail calls and to avoid switching the default calling convention only those marked will need to get special treatment.
5. Finally, just required general tail call support of the backends but we wouldn't having this discussion if it were that simple.

Author: > Are tail calls needed between different crates?

I can imagine libraries that you would want to tail call into for best performance. Also personally, I would be interested in tail calling functions that result from jitting, these functions I would categorize as external functions.

Reviewer 9: > However, what exactly would we win by having a constraint to only allow tail calls within the same crate if that's why you were asking?

The big thing is in separate compilation possibilities.  If tail calls are only guaranteed inside a crate, then it would be possible (if admittedly disruptive to how rustc currently works) to find the connected component of things that tail call each other, and compile them to one big CFG, with any exported functions calling into that one big merged feature in a way that starts at the right basic block.

That would, thus, give guaranteed tail calls entirely within rustc, without backends needing to support anything, because they'd only see a big CFG, not any *calls* at all.

Author: Correct me if I'm wrong but this sounds like option one mentioned above.

Reviewer 10: For the sake of completeness, there is another way to do this without backend support, but it's kind of ridiculous.

What you do is make normal calls, allowing the stack to expand, plus do some extra bookkeeping.  You emit code to catch the stack overflow.  When that happens, you use that extra bookkeeping to unwind the whole stack as though tail calls had been happening all along.

This is the approach that Chicken Scheme takes when compiling to standards-compliant C code.  It's called [Cheney on the M.T.A.].

[Cheney on the M.T.A.]: https://web.archive.org/web/20191231120238/http://home.pipeline.com/~hbaker1/CheneyMTA.html

Author: updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/dfcf7e84471c536fa75cbd4f2d8c4816e81dd503

Author: I'm resolving this thread as I think all points are now addressed. Please reopen if you disagree.

  - TCE as proposed in this RFC requires dropping local variables before the function call instead of after with `return`.
  - From a type system perspective the type of the `return` expression (`!`, the never type, see [here](https://doc.rust-lang.org/std/primitive.never.html) for an example) stays the same even when adding one of the markings. This means that type-checking can not help if the marking is forgotten or added mistakenly. (Note that the argument, the function call, can still be type checked, just not the `return` expression.)
   
### Require Explicit Dropping of Variables

(Based on this [comment](https://github.com/rust-lang/rfcs/pull/3407#issuecomment-1532841475))

An alternative approach could be to refuse to compile functions that would need to run destructors before `become`.
This would force code to not rely on implicit drops and require calls to `drop(variable)`, as in the following example:

```rust
fn f(x: String) {
    drop(x); // necessary
    become g();
}
```

This approach would result in more verbose code but would also be easier to read for people not familiar with tail calls.
Also, it is forwards compatible with implicit dropping before `become`.

> The reason this approach is not chosen is that the tradeoff between increased verbosity and the reduction of initial learning time seems to not be worth it.

Reviewer 2: I want to add that the implementation of the diagnostic for forgotten drops would likely be *harder*, than the implementation of correct drop elaboration. (this doesn't *necessarily* need to be added to the RFC, but just though I'd comment)

Author: added here: https://github.com/rust-lang/rfcs/pull/3407/commits/0e4a645a87243116fa9d84bf0f77047cc5b79df7

Added, just in case this alternative wasn't unappealing enough. :sweat_smile:

> Clang requires the argument types, argument number, and return type to be the same between the
> caller and the callee, as well as out-of-scope considerations such as C++ features and the calling
> convention. Implementor experience with Clang shows that the ABI of the caller and callee must be
> identical for the feature to work; otherwise, replacement may be impossible for some targets and
> conventions (replacing a differing argument list is non-trivial on some platforms).

## GCC
GCC does not support a feature equivalent to Clang's `musttail`, there also does not seem to be a push to implement it ([pipermail](https://gcc.gnu.org/pipermail/gcc/2021-April/235882.html)) (as of 2021). However, there also exists an experimental [plugin](https://github.com/pietro/gcc-musttail-plugin) for GCC last updated in 2021.

## Zig
Zig provides separate syntax to allow more flexibility than normal function calls. There are options for async calls, inlining, compile-time evaluation of the called function, or specifying TCE on the call.
([source](https://ziglang.org/documentation/master/#call))
```zig
const expect = @import("std").testing.expect;

test "noinline function call" {

>     try expect(@call(.auto, add, .{3, 9}) == 12);

Reviewer 2: Would be nicer to show an example with tail calls directly. The Zig documentation doesn't seem to provide one, but maybe someone knowledgeable with it can write one.

Author: Yes, it would be nice to have one. I didn't find an example in the documentation either. Also community sentiment is missing, as the TODO below indicates.

Author: updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/51ecc12222b02663da5a581875be8e8b87025959

At least a working example.

Author: Resolving as we now have an example, please reopen if you disagree.

    - Should a lint be added for functions that are marked to be a tail call or use become. See the discussion [here](https://github.com/rust-lang/rfcs/pull/3407#discussion_r1159822824), as well as, the clippy and rustfmt changes of an initial [implementation](https://github.com/semtexzv/rust/commit/29f430976542011d53e149650f8e6c7221545207#diff-6c8f5168858fed7066e1b6c8badaca8b4a033d0204007b3e3025bf7dd33fffcb) (2022).
- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?
    - Are all calling-convention used by Rust available for TCE with the proposed restrictions on function signatures?
    - Can the restrictions on function signatures be relaxed?
        - One option for intra-crate direct calls is to automatically pad the arguments during compilation see [here](https://github.com/rust-lang/rfcs/pull/3407#issuecomment-1500620309). Does this have an influence on other calls? How much implementation effort is it?

>     - Can generic functions be supported?

Reviewer 2: Given that Rust uses monomorphization, I don't see any reason why generic functions may not be supported.

Author: You are right, generic functions should not be a problem. Just for context the list here is of all features that I have found to be discussed in the context of tail calls. So if the answer to some of them is that there is no problem, then this was just me being overly cautious.

Reviewer 2: Thanks for being cautious! I think, however, that we should resolve the things we can, to make the list shorter and easier to grasp/focus on real questions.

Author: Agreed, just didn't get to it yet. Updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/eeaa80c6db06ab1e9a7ecb755844e8b3da7baef8

    - Can async functions be supported? (see [here](https://github.com/rust-lang/rfcs/pull/1888#issuecomment-1186604115) for an initial assessment)
    - Can closures be supported? (see [here](https://github.com/rust-lang/rfcs/pull/1888#issuecomment-1186604115) for an initial assessment)

>     - Can dynamic function calls be supported?

Reviewer 2: Similarly I haven't seen any problems with tail calling function pointers, given that they have the matching signature.
```suggestion
    - Can dynamic function calls be supported?
```

Author: resolving as this comment is a duplicate

>     - Can dynamic function calls be supported?

Reviewer 2: Similarly I haven't seen any problems with tail calling function pointers, given that they have the matching signature.
```suggestion
```

Author: Updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/eeaa80c6db06ab1e9a7ecb755844e8b3da7baef8

>     - Can dynamic function calls be supported?

Reviewer 2: Similarly I haven't seen any problems with tail calling function pointers, given that they have the matching signature.

Author: Updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/eeaa80c6db06ab1e9a7ecb755844e8b3da7baef8

>     - Can functions outside the current crate be supported, functions from dynamically loaded libraries?

Reviewer 2: Similarly to dynamic functions, as long as the signatures match, there is no problem (at least for the LLVM backend).

>     - Can functions that abort be supported?

Author: I was thinking of backtraces here, though I guess the tail called function's frames would just be omitted, except if we want to do something special here. I would be fine with removing this question.

>     - Can functions that abort be supported?

Reviewer 2: Yes, I expect backtraces to "work", modulo the fact that they'll be missing all the tail calls. (maybe this should be documented in the RFC?)

>     - Can functions that abort be supported?

Author: Sure, if it's sufficient, let resolve this question.

    - Is there some way to reduce the impact on debugging and other features?
- What related issues do you consider out of scope for this RFC that could be addressed in the future independently of
  the solution that comes out of this RFC?
  - Supporting general tail calls, the current RFC restricts function signatures which can be loosened independently in the future.

> # Future possibilities

Reviewer 2: I would like to see a future possibility of a special tail-call-specific calling convention that lifts some of the restrictions, as long as both function use this special calling convention (say `rusttail`).

Author: Just to make sure this would be similar to the alternative: "Attribute on Function Declaration"?

So the idea would be to allow using `become` for these marked functions, even if the signatures do not match exactly?

Author: Actually, I think that alternative is already close to what you describe: https://github.com/phi-go/rfcs/blob/guaranteed-tco/text/0000-explicit-tail-calls.md#attribute-on-function-declaration.

I would add that this alternative could be adopted in the future additionally to using `become`.

Author: Added here: https://github.com/rust-lang/rfcs/pull/3407/commits/3010c2e68c568f6c8741814c20034a5a38086c0e

Author: Added here: https://github.com/rust-lang/rfcs/pull/3407/commits/3010c2e68c568f6c8741814c20034a5a38086c0e

Author: I'm resolving this as I think it's adressed, please reopen if you disagree.

```rust
use std::panic::Location;

/// Returns the [`Location`] at which it is called.
#[track_caller]
fn get_caller_location() -> &'static Location<'static> {
    Location::caller()
}

/// Returns a [`Location`] from within this function's definition.
fn get_just_one_location() -> &'static Location<'static> {
    become get_caller_location();
}
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation
<!-- This is the technical portion of the RFC. Explain the design in sufficient detail that:

- Its interaction with other features is clear.
- It is reasonably clear how the feature would be implemented.
- Corner cases are dissected by example.

The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work. -->
Implementation of this feature requires checks that all prerequisites to guarantee TCE are fulfilled.
These checks are:

- The `become` keyword is only used in place of `return`. The intent is to reuse the semantics of a `return` signifying "the end of a function". See the section on [tail-call-elimination](#tail-call-elimination) for examples.

> - The argument to `become` is a function (or method) call, that exactly matches the function signature and calling convention of the callee. The intent is to ensure a matching ABI. Note that mutability and lifetimes may differ as long as they pass borrow checking.

Reviewer 2: Can you leave mutability as a future possibility/unresolved question? Allowing coercing mutability is somewhat annoying impl vise.

Author: updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/12c495b3a86326d0b4e0ece4bf5e72a04975757a

Added as an unresolved question as I think this needs a bit more thought on how important it is. While adding `mut` to all arguments seems similar to requiring that all arguments match, it feels like that approach clashes a bit with allowing variable lifetimes.

I would be interested the general sentiment on this!

Reviewer 11: I don't think mutability coercion should be allowed. My preferred way to express the "modulo lifetimes" issue is that the *function pointer* type of the called function must be coercible to the function pointer for the current function (or possibly vice versa, I didn't think too hard about the variance requirements here). That automatically ensures ABI compatibility while still allowing a lot of lifetime variation. Changing `&T` <-> `&mut T` is not allowed in this context, in either the arguments or the return.

Reviewer 2: > Added as an unresolved question as I think this needs a bit more thought on how important it is. While adding `mut` to all arguments seems similar to requiring that all arguments match, it feels like that approach clashes a bit with allowing variable lifetimes.

I'm not sure I understand... do you mean that the following is allowed:
```rust
// snippet 1
fn f(x: u8) { become g(x) }
fn g(mut x: u8) {}
```
Or this is allowed:
```rust
// snippet 2
fn f(x: &mut u8) { become g(x) }
fn g(x: &u8) {}
```

Snippet 1 seems uncontroversial, `mut` there is a property of the binding in the function body, not the function signature. Snippet 2 on the other hand is a lot more controversial, because it needs a guarantee that `fn(&mut u8)` and `fn(&u8)` are ABI compatible, which in practice seems to be true, but it's quite a thing to guarantee...

Author: I meant snippet 2, I agree that snippet 1 is uncontroversial. As for tail calls we only care about ABI compatibility it seemed like this mismatch might be possible to support. Though, it wasn't clear to me that it is not already guaranteed. I'll change it to a future possibility if this should ever be guaranteed.

Should we mention snippet 1 in the RFC? It seems obvious enough to not require a separate description.

Reviewer 2: I don't think we need to mention snippet 1 in the RFC.

Author: added here (forgot to add it to the commit message): https://github.com/rust-lang/rfcs/pull/3407/commits/3010c2e68c568f6c8741814c20034a5a38086c0e

Reviewer 3: > ```rust
> // snippet 2
> fn f(x: &mut u8) { become g(x) }
> fn g(x: &u8) {}
> ```

Doesn't auto-de/ref kick-in here? If it's desugared to `become g(&*x);` after tyck, it works for free according to this RFC.

Reviewer 11: @oxalica the issue is the "same arguments" restriction, since `f` and `g` do not have the same signature this would likely be disallowed. Autoref/deref in arguments isn't a problem, that would just be a normal coercion and there are no restrictions on operations in the function arguments.

Reviewer 12: Requiring `fn(&u8)` and `fn(&mut u8)` to have the same ABI would also preclude references to small types (without interior mutability) being passed by its content directly (other semantic issues notwithstanding).

Reviewer 4: That already isn't possible due to pointer addresses being observable. If you can prove that the pointer address won't be observed, you can also prove the absence of `become` and only then do the `&T` -> `T` transformation.

Reviewer 12: > That already isn't possible due to pointer addresses being observable.

I’m aware of that – that was the other semantic issue I was mentioning. Still, allowing snippet 2 above would make this already turbulent change even harder to introduce. Whether we should treat addressless references as a lost cause or determine that it has enough benefits to leave on the table is of course a matter of debate, and one that’s probably out of scope here.

Reviewer 13: We could leave snippet 2 unsupported (for now?) and it would have a straightforward workaround:

```
fn f(x: &mut u8) { 
    let y = &*x;
    become g(y) 
}

fn g(x: &u8) {}
```

Author: > That already isn't possible due to pointer addresses being observable. If you can prove that the pointer address won't be observed, you can also prove the absence of `become` and only then do the `&T` -> `T` transformation.

As someone unfamiliar with that optimization, is this optimization only planned for static function calls? Otherwise, I'm not sure how one would prove the absence of `become` as the current RFC supports tail calling dynamic calls (or I'm misunderstanding something).

Reviewer 4: If you can't prove that the address isn't observable (like in the case of a dynamic call), then you can't turn `&T` arguments into `T`. If you can prove that the address isn't observable, you can also prove the absence of `become`. As such the `&T` to `T` arg optimization shouldn't matter for `become` as the set of cases where you can prove that it is allowed fully contains the set of cases where `become` isn't used.

Reviewer 14: Let we have also Snippet 3 and 4:
```rust
// snippet 3
fn f(x: &mut u8) { become g(x) }
fn g(x: &u8) { if .. {become g(x)}}

// snippet 4
fn f(x: u8) { if .. {become g(x) }}
fn g(mut x: u8) { if .. {become f(x) }}
```
What to do here?

Author: @VitWW the content of a function can not change the ABI of a function, so both snippets are equivalent to snippet 2 regarding this discussion. The "workaround" would need to be that the programmer adapts the function signature.

The type of the expression `become <call>` is `!` (the never type, see [here](https://doc.rust-lang.org/std/primitive.never.html)). This is consistent with other control flow constructs such as `return`, which also have the type of `!`.

Note that as `become` is a keyword reserved for exactly the use-case described in this RFC there is no backward-compatibility break. This RFC only specifies the use of `become` inside of functions and instead leaves usage outside of functions unspecfied for use by other features.

This feature will have interactions with other features that depend on stack frames, for example, debugging and backtraces. See [drawbacks](#drawbacks) for further discussion.

See below for specifics on interations with other features.

## Coercions of the Tail Called Function's Return Type

All coercions that do any work (like deref coercion, unsize coercion, etc) are prohibited.
Lifetime-shortening coercions (`&'static T` -> `&'a T`) are allowed but will be checked by the borrow checker.

> Note that, while in theory, never-to-any coercions (`! -> T`) could be allowed, they are difficult to implement and require backend support. As a result they are not allowed as per this RFC. This has no effect on using macros like `panic!()` as they are not functions, affected are only functions like the following:

Author: As panic related macros also have the never type I wanted to mention that they can still be used in functions. I didn't mean that they can be tail called, though, I agree this is confusingly worded currently.

> Note that, while in theory, never-to-any coercions (`! -> T`) could be allowed, they are difficult to implement and require backend support. As a result they are not allowed as per this RFC. This has no effect on using macros like `panic!()` as they are not functions, affected are only functions like the following:

Author: also added pointer/reference coercions: https://github.com/rust-lang/rfcs/pull/3407/commits/c5301f2acac89c1d33904b70b374f3b3986fe193

```rust
fn never() -> ! {
    loop {}
}

fn tail_call_never_type() -> usize {
    become never();
}
```

## Closures
[closures]: #closures

Tail calling closures _and_ tail calling _from_ closures is **not** allowed.
This is due to the high implementation effort, see below, this restriction can be lifted by a future RFC.

Closures use the `rust-call` unstable calling convention, which would need to be adapted to guarantee TCE.
Additionally, any closure that has captures would need special handling, since the captures would currently be dropped before the tail call.

## Variadic functions using `c_variadic`

Tail calling [variadic functions](https://doc.rust-lang.org/beta/unstable-book/language-features/c-variadic.html) _and_ tail calling _from_ variadic functions is **not** allowed.
As support for variadic function is stabilized on a per target level, support for tail-calls regarding variadic functions would need to follow a similar approach. To avoid this complexity and to minimize implementation effort for backends, this interaction is currently not allowed but supported can be added with a future RFC.

## Generators

> Tail calling [generators](https://doc.rust-lang.org/beta/unstable-book/language-features/generators.html) is **not** allowed as it is a fundamental mismatch of functionality. Generators expect to be called multiple times `yield`ing values, however, when using a tail call control would never be returned to the calling function.

Reviewer 2: Not sure what is meant by this, you can't call a generator anyway (and tail-calling `resume` should be possible, it's just a method). I'd only leave the "from generators" restriction.

Author: updated: https://github.com/rust-lang/rfcs/pull/3407/commits/4b184876a5334f2e6f4b445e7cc590ee4cbf2137

Tail calling from generators is also **not** allowed, as the generator state is stored internally, tail calling from the generator function would require additional support to function correctly. To limit implementation effort this is not supported but can be supported by a future RFC.

## Async

> Tail calling async functions is **not** allowed as it requires special support for the async state machine.

Author: updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/b3dc340150cd6244baa91eff0220e6b66dcfee3c

Please give it another read, I'm not that familiar with async.

> Tail calling async functions is **not** allowed as it requires special support for the async state machine.

Reviewer 10: Shouldn't `become f().await` be the same as `become (f().await)`?  I.e. it should mean:

```rust
let _tmp = f().await;
become _tmp;
```

So if `f().await` returns a compatible function pointer then this would be legal.

> Tail calling async functions is **not** allowed as it requires special support for the async state machine.

Reviewer 2: I've written a few comments on the async section in the review down below.

These checks are:

- The `become` keyword is only used in place of `return`. The intent is to reuse the semantics of a `return` signifying "the end of a function". See the section on [tail-call-elimination](#tail-call-elimination) for examples.
- The argument to `become` is a function (or method) call, that exactly matches the function signature and calling convention of the callee. The intent is to ensure a matching ABI. Note that lifetimes may differ as long as they pass borrow checking, see [below](#return-type-coercion) for specifics on the return type.
- The stack frame of the calling function is reused, this also implies that the function is never returned to. The required checks to ensure this is possible are: no borrows of local variables are passed to the called function (passing local variables by copy/move is ok since that doesn't require the local variable to continue existing after the call), and no further cleanup is necessary. These checks can be done by using the borrow checker as already described in the [section](#difference) showing the difference between `return` and `become` above.

> 

Reviewer 2: I think we should add something like

> - Additionally there are some restrictions on the caller context/callee signature, see below

since we have more restrictions below, like banning `c_variadic` or explicitrly banning closures, etc.

Author: updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/0b61f079d8eb0a8caf216d1937a770d5fafb188e

If any of these checks fail a compiler error is issued.

One additional check must be done, if the backend cannot guarantee that TCE will be performed an ICE is issued. It is also suggested to ensure that the invariants provided by the pre-requisites are maintained during compilation, raising an ICE if this is not the case.

The type of the expression `become <call>` is `!` (the never type, see [here](https://doc.rust-lang.org/std/primitive.never.html)). This is consistent with other control flow constructs such as `return`, which also have the type of `!`.

Note that as `become` is a keyword reserved for exactly the use-case described in this RFC there is no backward-compatibility break. This RFC only specifies the use of `become` inside of functions and instead leaves usage outside of functions unspecfied for use by other features.

This feature will have interactions with other features that depend on stack frames, for example, debugging and backtraces. See [drawbacks](#drawbacks) for further discussion.

See below for specifics on interations with other features.

## Mismatches in Mutability

> Mismatches in mutability (like `&T` <-> `&mut T`) for arguments and return type of the function signatures are **not** supported. This support requires a guarantee that mutability has no effect on ABI.

Reviewer 2: Not sure if this paragraph is needed — `&T` and `&mut T` are different types, so there shouldn't be any questions regarding this. Maybe as a future possibility it would make more sense. Anyway if we'll keep it somewhere, "This support" should be changed to "Supporting this".

Author: updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/b7c69e4a21a5d9552a744fb02bcbab16d455b351

While unlikely to happen, I think having a note on this is still worthwhile.

Additionally, any closure that has captures would need special handling, since the captures would currently be dropped before the tail call.

## Variadic functions using `c_variadic`

Tail calling [variadic functions](https://doc.rust-lang.org/beta/unstable-book/language-features/c-variadic.html) _and_ tail calling _from_ variadic functions is **not** allowed.
As support for variadic function is stabilized on a per target level, support for tail-calls regarding variadic functions would need to follow a similar approach. To avoid this complexity and to minimize implementation effort for backends, this interaction is currently not allowed but supported can be added with a future RFC.

## Generators

Tail calling from [generators](https://doc.rust-lang.org/beta/unstable-book/language-features/generators.html) is **not** allowed. As the generator state is stored internally, tail calling from the generator function would require additional support to function correctly. To limit implementation effort this is not supported but can be supported by a future RFC.

## Async
[async]: #async

> Tail calling _from_ async functions is **not** allowed, neither calling async nor calling sync functions is supported. This is due to the high implementation effort as it requires special handling for the async state machine. This restriction can be relaxed by a future RFC.

Reviewer 2: ```suggestion
Tail calling _from_ async functions or async blocks is **not** allowed. This is due to the high implementation effort as it requires special handling for the async state machine. This restriction can be relaxed by a future RFC.
```

> Using `become` on a `.await` call, such as `become f().await`, is also **not** allowed. This is because when using `.await`, the `Future` returned by `f()` is not "called" but run by the executor, thus, tail calls do not apply here.

Reviewer 2: ```suggestion
Using `become` on a `.await` expression, such as `become f().await`, is also **not** allowed. This is because `become` requires a function call and `.await` is not a function call, but is a special construct.
```

> Note that tail calling async functions from sync code is possible but the return type for async functions is `std::future::Future`, which is unlikely to be interesting.

Reviewer 2: ```suggestion
Note that tail calling async functions from sync code is possible but the return type for async functions is `impl Future`, which is unlikely to be interesting.
```

## Operators are not supported

Invocations of operators were considered as valid targets but were rejected on grounds of being too error-prone.
In any case, these can still be called as methods. One example of their error-prone nature ([source](https://github.com/rust-lang/rfcs/pull/3407#discussion_r1167112296)):
```rust
pub fn fibonacci(n: u64) -> u64 {
    if n < 2 {
        return n
    }
    become fibonacci(n - 2) + fibonacci(n - 1)
}
```
In this case, a naive author might assume that this is going to be a stack space-efficient implementation since it uses tail recursion instead of normal recursion. However, the outcome is more or less the same since the critical recursive calls are not actually in tail call position.

> Further confusion could result from the same-signature restriction where the Rust compiler raises an error since fibonacci and <u64 as Add>::add do not share a common signature.

Reviewer 2: ```suggestion
Further confusion could result from the same-signature restriction where the Rust compiler raises an error since fibonacci and `<u64 as Add>::add` do not share a common signature.
```

>
> Just to be clear, the Edge team is against implicit tail-calls for all functions, but they're in favor of tail-calls-with-an-explicit-keyword (similar to this RFC).

An unofficial summary of the ECMA Script/ Javascript proposal for tail call/return
https://github.com/carbon-language/carbon-lang/issues/1761#issuecomment-1198672079 (Jul, 2022)

# Unresolved questions
[unresolved-questions]: #unresolved-questions
<!--
- What parts of the design do you expect to resolve through the RFC process before this gets merged?
- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?
- What related issues do you consider out of scope for this RFC that could be addressed in the future independently of the solution that comes out of this RFC? -->

- What parts of the design do you expect to resolve through the RFC process before this gets merged?
    - One point that needs to be decided is if TCE should be a feature that needs to be required from all backends or if it can be optional. Currently, the RFC specifies that an ICE should be issued if a backend cannot guarantee that TCE will be performed.

>     - Another point that needs to be decided is if TCE is supported by a backend what exactly should be guaranteed? While the guarantee that there is no stack growth should be necessary, should performance (as in transforming `call` instructions into `jmp`) also be guaranteed? Note that a backend that guarantees performance should do so **always** otherwise the main intent of this RFC seems to be lost.

Reviewer 2: My take on this is that we can't really guarantee performance, it is not really quantifiable and we don't have enough control to guarantee anything.

I think the LLVM's guarantee is the most reasonable we can get:
> The call will not cause unbounded stack growth if it is part of a recursive cycle in the call graph.

Author: Maybe I'm being too pedantic but can you explain why point 1 is enough? Point 1 alone would allow me to implement a backend that just does normal calls until the stack overflows and then clean up, as Chicken Scheme does, this approach does not have unbounded stack growth.

I think point 2 is close to forbidding new stack frames so it could be adopted by us. (Point 3 does not apply I agree.)

Author: Yeah, on second thought point `2.` does not really help, we could at most also require that references are passed in place. Which we should?

So I guess we can't require a "guaranteed performance". I would be interested what the opinions on this from a backend perspective are. I'm not even sure if this guarantee is just impossible to provide by a backend or that LLVM just wanted to have the freedom to do different things.

Author: Sure, so just keep it? If yes, just resolve. I think this question will need reaching out to someone more knowledgeable.

Reviewer 2: I did not understand your question 😅

Author: Ah, sorry. I meant should anything be changed in the RFC or just keep it as it is?

Reviewer 15: On further thought, I guess both, even if syntactically disambiguated, remain semantically ambiguous between
1. "if it is part of a recursive cycle in the call graph, the call will not cause unbounded stack growth [in the recursive cycle]" and
2. "if it is part of a recursive cycle in the call graph, the call will not cause unbounded stack growth [even on other call paths]".

Reviewer 11: I don't think there is any syntax ambiguity in that sentence. (2) follows the sentence structure and I have no idea how to interpret (1). Your rewrites to put the "if" part first in the second post look correct.

For the second ambiguity, it seems most reasonable to use interpretation (1), since if a tail-recursive cycle calls a non-tail-recursive cycle obviously nothing can be done about the stack growth that happens due to the second recursion even if the first one only shows up once in the call stack.

Author: updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/2946d6346048cd9737f4a7d73fd5bada0589ea44

Added as a future possibility, I hope that is fine.

way. Try to use this section as a tool to more fully consider all possible
interactions with the project and language in your proposal.
Also consider how this all fits into the roadmap for the project
and of the relevant sub-team.

This is also a good place to "dump ideas", if they are out of scope for the
RFC you are writing but otherwise related.

If you have tried and cannot think of any future possibilities,
you may simply state that you cannot think of anything.

Note that having something written down in the future-possibilities section
is not a reason to accept the current or a future RFC; such notes should be
in the section on motivation or rationale in this or subsequent RFCs.
The section merely provides additional information. -->

## Helpers
[helpers]: #helpers

It seems possible to keep the restriction on exactly matching function signatures by offering some kind of placeholder
arguments to pad out the differences. For example:

```rust
foo(a: u32, b: u32) {
    // uses `a` and `b`
}

bar(a: u32, _b: u32) {
    // only uses `a`
}
```

Maybe it is useful to provide a macro or attribute that inserts missing arguments.

```rust
#[pad_args(foo)]
bar(a: u32) {
    // ...
}

> ```

Reviewer 2: In those 2 examples `fn` tokens are missing

Author: whoops, updated: https://github.com/rust-lang/rfcs/pull/3407/commits/23106d2e678363ee4d4d6c6f16e535539dcd3e65

- Example [BPF code](https://blog.cloudflare.com/assembly-within-bpf-tail-calls-on-x86-and-arm/) from a Cloudflare blog post
- [RSM](https://github.com/rsms/rsm/blob/d539fd5f09876700c0c38758f2b4354df433dd1c/src/rsmimpl.h#L115) (a virtual computer in the form of a virtual machine)
- [Tails](https://github.com/snej/tails/blob/d3b14fcce18c542211bc1fd37e378f667fdee42f/src/core/platform.hh#L52) (a Forth-like interpreter)
- [Jasmin](https://github.com/asoffer/jasmin/blob/f035ef0752c09846331c8deb2109e4ebfce83200/jasmin/internal/attributes.h#L13) (a stack-based byte-code interpreter)
- [CHERIoT RTOS](https://github.com/microsoft/cheriot-rtos/blob/3e6811279fedd0195e105eb3b7ac77db93d67ec5/sdk/core/allocator/alloc.h#L1460) (a realtime operating system with memory safety)
- [Pyston](https://github.com/pyston/pyston/blob/6103fc013e9dd726efca9100a22be1ac08c58591/pyston/aot/aot_gen.py#L276) (a performance-optimizing JIT for Python)
- [upb](https://github.com/classicvalues/upb/blob/2effcce774ce05d08af635ba02b1733873e73757/upb/port_def.inc#L177) (a small protobuf implementation in C)

> - [wasm3](https://github.com/Robbepop/wasm3/blob/1a6ca56ee1250d95363424cc3a60f8fd14f24fa7/source/m3_config_platforms.h#L86) ("the self-proclaimed fastest WebAssembly interpreter")

Reviewer 2: @Robbepop are your fork any different from the `wasm3/wasm3`? it looks like the upstream has the same tail calls.

Reviewer 5: Ah good catch, I didn't even know that I forked this project. Please link to the original Wasm3 project!
https://github.com/wasm3/wasm3

Reviewer 10: I updated my comment to reference the upstream project rather than @Robbepop's fork.  This and the other changes should probably be copied.

> - [wasm3](https://github.com/Robbepop/wasm3/blob/1a6ca56ee1250d95363424cc3a60f8fd14f24fa7/source/m3_config_platforms.h#L86) ("the self-proclaimed fastest WebAssembly interpreter")

Reviewer 5: ```suggestion
- [wasm3](https://github.com/wasm3/wasm3/blob/1a6ca56ee1250d95363424cc3a60f8fd14f24fa7/source/m3_config_platforms.h#L86) ("the self-proclaimed fastest WebAssembly interpreter")
```

Reviewer 5: Please link to the original project instead of my no-op fork. 🙃

Never-to-any coercions (`! -> T`) of the return type are **not** supported to minimize implementation effort. They are difficult to implement and require backend support. To be clear, this only concerns functions that have the never return type like the following example:

```rust
fn never() -> ! {
    loop {}
}

fn tail_call_never_type() -> usize {
    become never(); //~ error: mismatched types
}
```

## Closures
[closures]: #closures

Tail calling closures _and_ tail calling _from_ closures is **not** allowed.
This is due to the high implementation effort, see below, this restriction can be lifted by a future RFC.

Closures use the `rust-call` unstable calling convention, which would need to be adapted to guarantee TCE.
Additionally, any closure that has captures would need special handling, since the captures would currently be dropped before the tail call.

## Variadic functions using `c_variadic`

Tail calling [variadic functions](https://doc.rust-lang.org/beta/unstable-book/language-features/c-variadic.html) _and_ tail calling _from_ variadic functions is **not** allowed.

> As support for variadic function is stabilized on a per target level, support for tail-calls regarding variadic functions would need to follow a similar approach. To avoid this complexity and to minimize implementation effort for backends, this interaction is currently not allowed but supported can be added with a future RFC.

Reviewer 16: ```suggestion
As support for variadic function is stabilized on a per target level, support for tail-calls regarding variadic functions would need to follow a similar approach. To avoid this complexity and to minimize implementation effort for backends, this interaction is currently not allowed but support can be added with a future RFC.
```

- What related issues do you consider out of scope for this RFC that could be addressed in the future independently of the solution that comes out of this RFC? -->

- What parts of the design do you expect to resolve through the RFC process before this gets merged?
    - One point that needs to be decided is if TCE should be a feature that needs to be required from all backends or if it can be optional. Currently, the RFC specifies that an ICE should be issued if a backend cannot guarantee that TCE will be performed.
- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?
    - Are all calling-convention used by Rust available for TCE with the proposed restrictions on function signatures?
    - Is there some way to reduce the impact on debugging and other features?
- What related issues do you consider out of scope for this RFC that could be addressed in the future independently of
  the solution that comes out of this RFC?
  - Supporting general tail calls, the current RFC restricts function signatures which can be loosened independently in the future.

## Resolved Questions

- Can generic functions be supported?

>   - As Rust uses Monomophization, generic functions are not a problem.

Reviewer 16: ```suggestion
  - As Rust uses Monomorphization, generic functions are not a problem.
```

# Drawbacks
[drawbacks]: #drawbacks
<!-- Why should we *not* do this? -->
As this feature should be mostly independent of other features the main drawback lies in the implementation and
maintenance effort. This feature adds a new keyword which will need to be implemented not only in Rust but also in other
tooling. The primary effort, however, lies in supporting this feature in the backends:
- LLVM supports a `musttail` marker to indicate that TCE should be performed [docs](https://llvm.org/docs/LangRef.html#id327). Clang which already depends on this feature seems to only generate correct code for the x86 backend [source](https://github.com/rust-lang/rfcs/issues/2691#issuecomment-1490009983) (as of 2023-03-30).
- GCC does seem to support an equivalent `musttail` marker, though it is only accessible via the [libgccjit API](https://gcc.gnu.org/onlinedocs/gcc-7.3.0/jit/topics/expressions.html#gcc_jit_rvalue_set_bool_require_tail_call) ([source](https://github.com/rust-lang/rfcs/pull/3407#discussion_r1160013809)).

> - For WebAssembly see the discussion [here](https://github.com/rust-lang/rfcs/pull/3407#discussion_r1186003262). While general tail calls seem far off, supporting only matching function signatures seems more feasible. Also note that WebAssembly accepted tail calls into the [standard](https://github.com/WebAssembly/proposals/pull/157/) and Cranelift is now [working](https://github.com/bytecodealliance/rfcs/pull/29) towards supporting it.

Author: Thank you! Updated for WebAssembly and GCC: https://github.com/rust-lang/rfcs/pull/3407/commits/3304fee7542a56e3c671b8ce8a85070904cbf2ba

> - For WebAssembly see the discussion [here](https://github.com/rust-lang/rfcs/pull/3407#discussion_r1186003262). While general tail calls seem far off, supporting only matching function signatures seems more feasible. Also note that WebAssembly accepted tail calls into the [standard](https://github.com/WebAssembly/proposals/pull/157/) and Cranelift is now [working](https://github.com/bytecodealliance/rfcs/pull/29) towards supporting it.

Reviewer 17: Note that `tailcall` is currently not enabled in the default list of target features for the `wasm32-unknown-unknown` target: https://doc.rust-lang.org/beta/rustc/platform-support/wasm32-unknown-unknown.html#enabled-webassembly-features

> - For WebAssembly see the discussion [here](https://github.com/rust-lang/rfcs/pull/3407#discussion_r1186003262). While general tail calls seem far off, supporting only matching function signatures seems more feasible. Also note that WebAssembly accepted tail calls into the [standard](https://github.com/WebAssembly/proposals/pull/157/) and Cranelift is now [working](https://github.com/bytecodealliance/rfcs/pull/29) towards supporting it.

Author: Updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/709c9b933eeb148dc53e2b1b44455611891538c2

For simple tail recursion on an iterable (successors)[https://doc.rust-lang.org/stable/core/iter/fn.successors.html] can be used to compute a result for each element.
Nearly equivalently, a (combinator)[https://users.rust-lang.org/t/when-will-rust-have-tco-tce/20790/3] can be used to express a tail recursive function, however, allowing more flexibility regarding the returned result.

These approaches do not provide a way to express general tail calls, so do not fulfill a basic requirement we would like to achieve.

### Trampoline based Approach
There could be a trampoline-based approach
([comment](https://github.com/rust-lang/rfcs/pull/1888#issuecomment-326952763)) that can fulfill the semantic guarantee
of using constant stack space, though they can not be used to achieve the performance that the chosen design is capable
of.

Similarly, as mentioned [here](https://github.com/rust-lang/rfcs/pull/3407#discussion_r1190464739), an approach used by Chicken Scheme is to do normal calls and handle stack overflows by cleaning up the stack.

### Principled Local Goto
One alternative would be to support some kind of local goto natively, indeed there exists a

> [pre-RFC](https://internals.rust-lang.org/t/pre-rfc-safe-goto-with-value/14470/9?u=scottmcm) ([comment](https://github.com/rust-lang/rfcs/issues/2691#issuecomment-1458604986)) or another (LLVM specific idea)[https://internals.rust-lang.org/t/idea-for-safe-computed-goto-using-enums/21787?u=programmerjake]. These designs should be able to achieve the targeted performance and stack usage, though they seems to be quite difficult to implement or are backend specific. Also they do not seem to be as flexible as the chosen design, especially implementing tail calls to indirect or external functions do not seem to be feasible.

Reviewer 18: Note that the computed goto idea I proposed isn't LLVM specific, since computed goto originally was a gcc extension for C/C++ before LLVM started, so it should work just fine in gcc too.

Reviewer 18: imo computed goto is more useful than tail calls in some situations (e.g. better register allocation due to not being restricted by the function call ABI) and less useful in others, so I think both tail calls and computed goto are worth adding to rust.

Reviewer 19: There's nothing preventing an optimizer from doing the same sort register allocation with tail calls, if the callees are equivalently local to the goto targets.

And on the other hand, in practice tail calls using a fixed ABI can produce better register allocation than computed goto, because a big irreducible CFG can trip up the allocator's heuristics in ways that separate functions can be written to avoid.

The better argument for computed goto is to unlock intraprocedural borrow checking, but even then it should probably look more like a tail call to a local function (with parameters), analogous to break-with-value, and less like the C extension, which doesn't play well with block scope or destructors.

Reviewer 7: What about the following Lisp-inspired syntax?

```rust
tagbody! {
   'a: {  /* do some work */ goto 'b; }
   'b: { /* do some work */ 1 }
}
```

Reviewer 18: > What about the following Lisp-inspired syntax?
> 
> ```rust
> tagbody! {
>    'a: {  /* do some work */ goto 'b; }
>    'b: { /* do some work */ 1 }
> }
> ```

the [syntax I had proposed](https://internals.rust-lang.org/t/idea-for-safe-computed-goto-using-enums/21787?u=programmerjake) is essentially equivalent to that.

Reviewer 20: Personally, I would prefer something that looks more like a `match` statement.

Reviewer 10: - https://github.com/rust-lang/rfcs/pull/3720

Author: @programmerjake Would you agree that local gotos require backend support? Then I would update that section to mention that instead.

I would also add some text that local gotos could have other applications than for tail calls. While this seems to already be disputed I guess we do not really know until this is tried?

Author: > yes, assuming you meant computed goto (jumping to an address in a variable instead of to a known label), but afaik all targets on gcc and/or llvm already support computed goto -- the main issue i raised in that internals thread is LLVM doesn't have a way to refer to a label's address from another codegen unit, so that needs to be fixed or worked around if we're using enums and not just a raw pointer to hold the addresses of our labels.

Ah, thanks for the clarification. So for a known label this is in theory possible without backend support?

> computed gotos do not let you make tail calls, they only work within a function instead of calling between functions. they can just be used for similar applications such as making fast interpreters.

Yeah, sorry I meant other use cases than being a possible alternative to the functionality provided by tail calls.

Reviewer 4: > Ah, thanks for the clarification. So for a known label this is in theory possible without backend support?

Yes, MIR is a CFG already. Jumping to a known label uses the exact same MIR block terminator as is used for eg control flow converging back to the same code path after an if else. Only the MIR building code and steps before it would need to be updated.

Author: Updated in: https://github.com/rust-lang/rfcs/pull/3407/commits/45259067a31796e9292c7d04c0b461ce5f1a22aa

```rust
fn foo(x: i32) -> i32 {
    if x % 2 {
        let x = x / 2;
        // one branch uses `become`
        become foo(x);
    } else {
        let x = x + 3;
        // the other does not
        foo(x) // == return foo(x);
    }
}
```

### Pitfall 3: Alternating `become` and `return` calls still grows the stack.
([original example](https://github.com/rust-lang/rfcs/pull/1888#issuecomment-279062656))

Here one function uses `become` the other `return`, this is another potential source of confusion. This mutual recursion
would eventually overflow the stack. As mutual recursion can also happen across more functions, `become` needs to be
used consistently in all functions if TCE is desired. (Maybe it is also possible to create a lint for these
use cases as well.)

```rust
fn foo(n: i32) {
    // oops, we forgot become ..
    return bar(n); // or alternatively: `bar(n)`
}

fn bar(n: i32) {
    become foo(n);
}
```

### Pitfall 4: Caller location aka [`panic::Location::caller`](https://doc.rust-lang.org/std/panic/struct.Location.html#method.caller) and `#[track_caller]`

> Since `#[track_caller]` adds an argument, `#[track_caller]` functions can only tail call other `#[track_caller]` functions. So the following example will not compile.

Reviewer 17: Removing `#[track_caller]` from a function should not be a breaking change. I think the simplest option is to simply forbid `become` inside a `#[track_caller]` function.

Alternatively, we could define the `#[track_caller]` ABI in a way that allows the tail call to happen despite the mismatch. But that would be more complicated to implement, and could have performance implications.

Tail-calling to a  `#[track_caller]` function from a non-tracking function could also cause issues: the call would have to go through a shim, which might grow the stack, defeating the purpose of tail calling.

Author: Thank you, updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/33c76cdc353a1af8f29d72696f771305862d2581

```rust
pub fn fibonacci(n: u64) -> u64 {
    if n < 2 {
        return n
    }
    become fibonacci(n - 2) + fibonacci(n - 1)
}
```
In this case, a naive author might assume that this is going to be a stack space-efficient implementation since it uses tail recursion instead of normal recursion. However, the outcome is more or less the same since the critical recursive calls are not actually in tail call position.

Further confusion could result from the same-signature restriction where the Rust compiler raises an error since fibonacci and `<u64 as Add>::add` do not share a common signature.

### Caller location ([`panic::Location::caller`](https://doc.rust-lang.org/std/panic/struct.Location.html#method.caller) and `#[track_caller]`) are not supported

Since `#[track_caller]` adds an argument, `#[track_caller]` functions can only tail call other `#[track_caller]` functions. So the following example would not compile.
However, removing `#[track_caller]` should not be a breaking change, as [it is not part of the stable API surface](https://github.com/rust-lang/rust/issues/87401#issuecomment-910628405) for Rust.

> For tail calls, however, removing an argument is a breaking change, so functions with `#[track_caller]` are **not** allowed to use `become`. ([Though, loosening this restriction might be possible.](https://github.com/rust-lang/rfcs/pull/3407/files/3304fee7542a56e3c671b8ce8a85070904cbf2ba#r2202286180)). Additionally, tail calling into a `#[track_caller]` function from a non-tracking function [might require growing the stack](https://github.com/rust-lang/rfcs/pull/3407/files/3304fee7542a56e3c671b8ce8a85070904cbf2ba#r2202286180), as this would defeat the purpose of tail calling, this is also **not** allowed.

Author: Just to make sure, you mean adding `track_caller` to a tail-called function should not be breaking? So the equivalent of adding `track_caller`?

If that is what you meant, I'm a bit confused. On the one hand I can see that we want `track_caller` to be usable transparently but on the other hand tail calls are pretty much incompatible with the idea of tracking the caller anyway. Ignoring the matching function signature issue, what caller would you want tracked for a tail called function? So should this just be added as another action that is not allowed? Or should we move `track_caller` into the "Unresolved questions" section as we need to figure out how this needs to work?

> For tail calls, however, removing an argument is a breaking change, so functions with `#[track_caller]` are **not** allowed to use `become`. ([Though, loosening this restriction might be possible.](https://github.com/rust-lang/rfcs/pull/3407/files/3304fee7542a56e3c671b8ce8a85070904cbf2ba#r2202286180)). Additionally, tail calling into a `#[track_caller]` function from a non-tracking function [might require growing the stack](https://github.com/rust-lang/rfcs/pull/3407/files/3304fee7542a56e3c671b8ce8a85070904cbf2ba#r2202286180), as this would defeat the purpose of tail calling, this is also **not** allowed.

Reviewer 17: > what caller would you want tracked for a tail called function

The call would have to go through the shim that is also used for function pointers, so no caller would be tracked.

> For tail calls, however, removing an argument is a breaking change, so functions with `#[track_caller]` are **not** allowed to use `become`. ([Though, loosening this restriction might be possible.](https://github.com/rust-lang/rfcs/pull/3407/files/3304fee7542a56e3c671b8ce8a85070904cbf2ba#r2202286180)). Additionally, tail calling into a `#[track_caller]` function from a non-tracking function [might require growing the stack](https://github.com/rust-lang/rfcs/pull/3407/files/3304fee7542a56e3c671b8ce8a85070904cbf2ba#r2202286180), as this would defeat the purpose of tail calling, this is also **not** allowed.

Reviewer: > > what caller would you want tracked for a tail called function
> 
> The call would have to go through the shim that is also used for function pointers, so no caller would be tracked.

The shim however is not going to be tail-calling the actual fn implementation. This means that if I have a track-caller and a non-tracker-caller and have them mutually tail-call each other, the stack usage would grow.

- Example [BPF code](https://blog.cloudflare.com/assembly-within-bpf-tail-calls-on-x86-and-arm/) from a Cloudflare blog post
- [RSM](https://github.com/rsms/rsm/blob/d539fd5f09876700c0c38758f2b4354df433dd1c/src/rsmimpl.h#L115) (a virtual computer in the form of a virtual machine)
- [Tails](https://github.com/snej/tails/blob/d3b14fcce18c542211bc1fd37e378f667fdee42f/src/core/platform.hh#L52) (a Forth-like interpreter)
- [Jasmin](https://github.com/asoffer/jasmin/blob/f035ef0752c09846331c8deb2109e4ebfce83200/jasmin/internal/attributes.h#L13) (a stack-based byte-code interpreter)
- [CHERIoT RTOS](https://github.com/microsoft/cheriot-rtos/blob/3e6811279fedd0195e105eb3b7ac77db93d67ec5/sdk/core/allocator/alloc.h#L1460) (a realtime operating system with memory safety)
- [Pyston](https://github.com/pyston/pyston/blob/6103fc013e9dd726efca9100a22be1ac08c58591/pyston/aot/aot_gen.py#L276) (a performance-optimizing JIT for Python)
- [upb](https://github.com/classicvalues/upb/blob/2effcce774ce05d08af635ba02b1733873e73757/upb/port_def.inc#L177) (a small protobuf implementation in C)

> - Robbepop's [wasm3](https://github.com/Robbepop/wasm3/blob/1a6ca56ee1250d95363424cc3a60f8fd14f24fa7/source/m3_config_platforms.h#L86) ("the fastest WebASsembly interpreter")

Reviewer 5: ```suggestion
- [wasm3](https://github.com/Robbepop/wasm3/blob/1a6ca56ee1250d95363424cc3a60f8fd14f24fa7/source/m3_config_platforms.h#L86) ("the self-proclaimed fastest WebAssembly interpreter")
```

Reviewer 10: I updated the language in my comment to be more generic here.  Should probably just copy in all of the updates.

- Corner cases are dissected by example.

The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work. -->
Implementation of this feature requires checks that all prerequisites to guarantee TCE are fulfilled.
These checks are:
- The `become` keyword is only used in place of `return`. The intend is to reuse the semantics of a `return` signifying "the end of a function". See the section on [tail-call-elimination](#tail-call-elimination) for examples.
- The argument to `become` is a function (or method) call, that exactly matches the function signature and calling convention of the callee. The intend is to assure a compatible stack frame layout.

> - The stack frame of the calling function is reused, this also implies that the function is never returned to. The required checks to ensure this is possible are: no local variables are passed to the called function, and no further cleanup is necessary. These checks can be done by using the borrowchecker as described in the [Borrowchecking](#semantics) section below.

Reviewer 18: ```suggestion
- The stack frame of the calling function is reused, this also implies that the function is never returned to. The required checks to ensure this is possible are: no borrows of local variables are passed to the called function (passing local variables by copy/move is ok, since that doesn't require the local variable to continue existing after the call), and no further cleanup is necessary. These checks can be done by using the borrowchecker as described in the [Borrowchecking](#semantics) section below.
```

Author: Ah yes, I forgot to mention that part. Thanks for taking another look!

There are some designs that either can not achieve the same performance or functionality as the chosen approach. Though most other designs evolve around how to mark what should be a tail-call or marking what functions can be tail called. There is also the possibility of providing support for a custom backend (e.g. LLVM) or MIR pass.

### Rust Built-in Functionality

> For simple tail recursion on an iterable (successors)[https://doc.rust-lang.org/stable/core/iter/fn.successors.html] can be used to compute a result for each element.

Reviewer 16: ```suggestion
For simple tail recursion on an iterable, [`successors`](https://doc.rust-lang.org/stable/core/iter/fn.successors.html) can be used to compute a result for each element.
```

### Principled Local Goto
One alternative would be to support some kind of local goto natively, indeed there exists a

> [pre-RFC](https://internals.rust-lang.org/t/pre-rfc-safe-goto-with-value/14470/9?u=scottmcm) ([comment](https://github.com/rust-lang/rfcs/issues/2691#issuecomment-1458604986)) or another (LLVM specific idea)[https://internals.rust-lang.org/t/idea-for-safe-computed-goto-using-enums/21787?u=programmerjake]. These designs should be able to achieve the targeted performance and stack usage, though they seems to be quite difficult to implement or are backend specific. Also they do not seem to be as flexible as the chosen design, especially implementing tail calls to indirect or external functions do not seem to be feasible.

Reviewer 16: ```suggestion
[pre-RFC](https://internals.rust-lang.org/t/pre-rfc-safe-goto-with-value/14470/9?u=scottmcm) ([comment](https://github.com/rust-lang/rfcs/issues/2691#issuecomment-1458604986)) or another (LLVM specific idea)[https://internals.rust-lang.org/t/idea-for-safe-computed-goto-using-enums/21787?u=programmerjake]. These designs should be able to achieve the targeted performance and stack usage, though they seem to be quite difficult to implement or are backend specific. Also they do not seem to be as flexible as the chosen design, especially implementing tail calls to indirect or external functions do not seem to be feasible.
```

Author: @Rudxain thank you!

    - Are all calling-convention used by Rust available for TCE with the proposed restrictions on function signatures?
    - Can the restrictions on function signatures be relaxed?
        - One option for intra-crate direct calls is to automatically pad the arguments during compilation see [here](https://github.com/rust-lang/rfcs/pull/3407#issuecomment-1500620309). Does this have an influence on other calls? How much implementation effort is it?
    - Can generic functions be supported?
    - Can async functions be supported? (see [here](https://github.com/rust-lang/rfcs/pull/1888#issuecomment-1186604115) for an initial assessment)
    - Can closures be supported? (see [here](https://github.com/rust-lang/rfcs/pull/1888#issuecomment-1186604115) for an initial assessment)

>     - Can dynamic function calls be supported?

Author: updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/18ec62d04e1ed591323aa334c020f3dfc4fa0998
